### PR TITLE
Data layout optimization (strategy generation). Part 1:  strategy class with persistence

### PR DESCRIPTION
**Pull request by teamurko**
## Summary
This is part 1 of a new feature: data layout optimization library, strategy generation.
Added strategy class and persistence utilities. Refactored existing compaction app to use the library config.

The following 3 components will be added eventually:
1) DLO library that has primitives for generating data layout optimization strategies
2) App that generates strategies for all tables
3) Scheduling of the app

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [x] Refactoring
- [x] Documentation
- [ ] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

- [ ] Manually Tested on local docker setup. Please include commands ran, and their output.
- [x] Added new tests for the changes made.
- [ ] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [x] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.commons.lang.StringEscapeUtils;
+import org.apache.spark.sql.SparkSession;
+
+/** Utility class for strategies persistence. */
+public final class Utils {
+  public static final String DATA_LAYOUT_PROPERTY_KEY = "data-layout-strategies";
+
+  private Utils() {}
+
+  public static void saveStrategies(
```
Comment (rohitkum2506) [inline libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:19]: Is this meant to be used for test purposes only? 
I'm not clear on how the strategy generation and saving fits in DLO/compaction flow. Can you add some details pls?

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import com.linkedin.openhouse.datalayout.layoutselection.DataLayoutOptimizationStrategy;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.lang.StringEscapeUtils;
+import org.apache.spark.sql.SparkSession;
+
+/** Utility class for strategies persistence. */
+public final class Utils {
+  public static final String DATA_LAYOUT_PROPERTY_KEY = "data-layout-strategies";
```
Comment (rohitkum2506) [inline libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:15]: Should we follow format "compaction.data-layout-strategies", if this property is specific to compaction processes?
Easier to identify with grouping.

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import com.google.gson.reflect.TypeToken;
+import com.linkedin.openhouse.datalayout.layoutselection.DataLayoutOptimizationStrategy;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.lang.StringEscapeUtils;
+import org.apache.spark.sql.SparkSession;
+
+/** Utility class for strategies persistence. */
+public final class Utils {
```
Comment (sumedhsakdeo) [inline libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:14]: Utils? :) 
Seems like a Strategy DTO/DAO class.

**Review by sumedhsakdeo — COMMENTED** [review]
Just one comment, on a high level this looks great, I plan to review part 2 and part 3 and come back here.

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java (hunk 1 above)
Comment (abhisheknath2011) [inline libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:19]: Are we going to store DLO strategy and config in the corresponding table properties? It seems it is part of table properties. Or is there a plan to maintain them separately?

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java (hunk 3 above)
Comment (abhisheknath2011) [reply libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:14]: +1 for specific naming.

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java (hunk 2 above)
Comment (teamurko) [reply libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:15]: thanks, compaction is too specific, we can make it consistent with iceberg naming convention, like write.data-layout.strategies.

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java (hunk 3 above)
Comment (teamurko) [reply libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:14]: thanks, introduced dao class

File: libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java (hunk 1 above)
Comment (teamurko) [reply libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/persistence/Utils.java:19]: Yes, the intention is to keep strategies in a table property

**Review by sumedhsakdeo — APPROVED** [review]
Neat!

**Pull request merged**